modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -44,80 +53,134 @@ public KafkaContainer(final DockerImageName dockerImageName) {
-        withEnv("KAFKA_LOG_FLUSH_INTERVAL_MESSAGES", Long.MAX_VALUE + "");
-        withEnv("KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS", "0");
+        withCreateContainerCmdModifier(cmd -> {
+            cmd.withEntrypoint("sh");
+        });
+        withCommand("-c", "while [ ! -f " + STARTER_SCRIPT + " ]; do sleep 0.1; done; " + STARTER_SCRIPT);
     }
 
     public KafkaContainer withEmbeddedZookeeper() {
+        if (this.kraftEnabled) {
+            throw new IllegalStateException("Cannot configure Zookeeper when using Kraft mode");
+        }
         externalZookeeperConnect = null;
         return self();
     }
 
     public KafkaContainer withExternalZookeeper(String connectString) {
+        if (this.kraftEnabled) {
+            throw new IllegalStateException("Cannot configure Zookeeper when using Kraft mode");
+        }
         externalZookeeperConnect = connectString;
         return self();
     }
 
+    public KafkaContainer withKraft() {
+        if (this.externalZookeeperConnect != null) {
+            throw new IllegalStateException("Cannot configure Kraft mode when Zookeeper configured");
+        }
+        String actualVersion = DockerImageName.parse(getDockerImageName()).getVersionPart();
+        if (new ComparableVersion(actualVersion).isLessThan(MIN_KRAFT_TAG)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Provided Confluent Platform's version %s is not supported in Kraft mode (must be %s or above)",
+                    actualVersion,
+                    MIN_KRAFT_TAG
+                )
+            );
+        }
+        kraftEnabled = true;
+        return self();
+    }
+
     public String getBootstrapServers() {
         return String.format("PLAINTEXT://%s:%s", getHost(), getMappedPort(KAFKA_PORT));
     }
 
     @Override
     protected void configure() {
-        withEnv(
-            "KAFKA_ADVERTISED_LISTENERS",
-            String.format("BROKER://%s:9092", getNetwork() != null ? getNetworkAliases().get(0) : "localhost")
-        );
+        if (kraftEnabled) {
+            waitingFor(Wait.forLogMessage(".*Transitioning from RECOVERY to RUNNING.*", 1));
+        } else {
+            waitingFor(Wait.forLogMessage(".*\\[KafkaServer id=\\d+\\] started.*", 1));
+        }
+        // Use two listeners with different names, it will force Kafka to communicate with itself via internal
+        // listener when KAFKA_INTER_BROKER_LISTENER_NAME is set, otherwise Kafka will try to use the advertised listener
+        withEnv("KAFKA_LISTENERS", "PLAINTEXT://0.0.0.0:" + KAFKA_PORT + ",BROKER://0.0.0.0:9092");
+        withEnv("KAFKA_LISTENER_SECURITY_PROTOCOL_MAP", "BROKER:PLAINTEXT,PLAINTEXT:PLAINTEXT");
+        withEnv("KAFKA_INTER_BROKER_LISTENER_NAME", "BROKER");
 
-        String command = "#!/bin/bash\n";
-        if (externalZookeeperConnect != null) {
+        withEnv("KAFKA_BROKER_ID", "1");
```

Comment:
kafka cluster example is failing because the broker id already exists. See [broker.id](https://docs.confluent.io/platform/current/installation/configuration/broker-configs.html#brokerconfigs_broker.id) 

I delete it locally and all was running ok.

```suggestion
```

Reply:
Indeed, I moved most of the broker configuration to the `configure` method to focus the constructor on the container lifecycle but looks like it was a bad idea...

I just pushed a fixed version, let's wait for the CI but should be good